{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f861cb88",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import SeqIO \n",
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90408884",
   "metadata": {},
   "source": [
    "# Import the sequence(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c898d5b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCTCACATACCAGATGCGAATTCCCGAG', SingleLetterAlphabet())"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_record = SeqIO.read('sequences/test/sequence_test.fa','fasta')\n",
    "#seq_record.id \n",
    "seq_record.seq "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b2ee103",
   "metadata": {},
   "source": [
    "## Get the complement sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b251f7c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complement(seq):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A','a': 't', 'c': 'g', 'g': 'c', 't': 'a'} \n",
    "    bases = list(seq) \n",
    "    bases = [complement[base] for base in bases] \n",
    "    return ''.join(bases)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e384f297",
   "metadata": {},
   "source": [
    "## Multiple sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6f6cdd51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\n"
     ]
    }
   ],
   "source": [
    "sequ = list(SeqIO.parse('sequences/pathogens/First_batch.fa','fasta'))\n",
    "sequ[0].seq\n",
    "\n",
    "seq = []\n",
    "\n",
    "for i in sequ:\n",
    "    seq.append(complement(i.seq))\n",
    "    \n",
    "print(len(sequ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "30131da9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID                             G_03\n",
       "Seq    gggtagaatttcaggtgtagcggtgaaa\n",
       "Name: 2, dtype: object"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame()\n",
    "\n",
    "df['ID'] = [sequ[i].id for i in range(len(seq))]\n",
    "df['Seq'] = [seq[i] for i in range(len(seq))]\n",
    "\n",
    "df.iloc[2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0431a594",
   "metadata": {},
   "source": [
    "# Call blast from the NCBI website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "38161649",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trying sequence number 0...\n",
      "Done with sequence 0 !\n",
      "Trying sequence number 1...\n",
      "Done with sequence 1 !\n",
      "Trying sequence number 2...\n",
      "Done with sequence 2 !\n",
      "Trying sequence number 3...\n",
      "Done with sequence 3 !\n",
      "Trying sequence number 4...\n",
      "Done with sequence 4 !\n",
      "Trying sequence number 5...\n",
      "Done with sequence 5 !\n",
      "Trying sequence number 6...\n",
      "Done with sequence 6 !\n",
      "Trying sequence number 7...\n",
      "Done with sequence 7 !\n",
      "Trying sequence number 8...\n",
      "Done with sequence 8 !\n",
      "Trying sequence number 9...\n",
      "Done with sequence 9 !\n",
      "Trying sequence number 10...\n",
      "Done with sequence 10 !\n",
      "Trying sequence number 11...\n",
      "Done with sequence 11 !\n",
      "Trying sequence number 12...\n",
      "Done with sequence 12 !\n",
      "Trying sequence number 13...\n",
      "Done with sequence 13 !\n",
      "Trying sequence number 14...\n",
      "Done with sequence 14 !\n"
     ]
    }
   ],
   "source": [
    "result_handle = []\n",
    "\n",
    "Entrez.email = 'ness.louafi@gmail.com'\n",
    "\n",
    "for i in range(len(seq)):\n",
    "    print('Trying sequence number ' + str(i)+'...')\n",
    "    result_handle.append(NCBIWWW.qblast(\"blastn\", \"nt\", seq[i], hitlist_size = 20,expect=0.05))\n",
    "    print('Done with sequence ' + str(i) + ' !')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c54a088",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "050fb139",
   "metadata": {},
   "outputs": [],
   "source": [
    "NCBIWWW.qblast?\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0caf528d",
   "metadata": {},
   "source": [
    "# Display the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "0fdf2faf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "G_01\n",
      "match: gi|1858882544|gb|CP053794.1| Vibrio cholerae strain W10G chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1858878921|gb|CP053796.1| Vibrio cholerae strain SL4G chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1858875361|gb|CP053798.1| Vibrio cholerae strain SL5Y chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1858871712|gb|CP053800.1| Vibrio cholerae strain SO5Y chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1858867049|gb|CP053802.1| Vibrio cholerae strain L6G chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1858862310|gb|CP053804.1| Vibrio cholerae strain SL6Y chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2221334104|gb|CP064379.1| Vibrio furnissii strain PartQ-Vfurnissii-RM8376 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2217779589|dbj|AP023385.1| Vibrio cholerae O96 RIMD 2214334 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217775976|dbj|AP023383.1| Vibrio cholerae O77 RIMD 2214315 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217771038|dbj|AP023381.1| Vibrio cholerae O63 RIMD 2214301 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217766341|dbj|AP023379.1| Vibrio cholerae O51 RIMD 2214289 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217761885|dbj|AP023377.1| Vibrio cholerae O5 RIMD 2214243 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217758102|dbj|AP023375.1| Vibrio cholerae O45 RIMD 2214283 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217752797|dbj|AP023373.1| Vibrio cholerae O207 RIMD 2214445 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217747447|dbj|AP023371.1| Vibrio cholerae O17 RIMD 2214255 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217743904|dbj|AP023369.1| Vibrio cholerae O162 RIMD 2214400 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2217738609|dbj|AP023331.1| Vibrio cholerae O102 RIMD 2214340 DNA, chromosome 1, complete genome \n",
      "e value: 0.0264252\n",
      "match: gi|2215582720|dbj|AP024967.1| Vibrio cholerae V130003 DNA, chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2210577337|gb|CP093426.1| Vibrio cholerae O1 biovar El Tor strain Inaba chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2176109310|gb|CP090387.1| Vibrio cholerae strain GXFL1-4 chromosome 2, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2176086866|gb|CP090399.1| Vibrio cholerae strain PanChol chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2168781201|gb|CP089603.1| Vibrio furnissii strain VFN3 chromosome I, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2077954350|gb|CP077197.1| Vibrio cholerae strain PS4 chromosome I, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2077325866|gb|CP080462.1| Vibrio cholerae strain ICDC-VC702 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2077325863|gb|CP080465.1| Vibrio cholerae strain ICDC-VC3741 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2069588885|gb|CP078727.1| Vibrio cholerae strain Colony11 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2069588881|gb|CP078725.1| Vibrio cholerae strain Colony605 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2069588879|gb|CP078719.1| Vibrio cholerae strain Colony98 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2069588877|gb|CP078723.1| Vibrio cholerae strain Colony268 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2067522135|gb|CP076643.1| Vibrio sp. OG9-811 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2058661182|gb|CP077426.1| Vibrio mimicus strain Y4 chromosome II, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2046371358|dbj|AP024555.1| Vibrio cholerae IDH-03506 DNA, chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2046371356|dbj|AP024553.1| Vibrio cholerae IDH-03329 DNA, chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2046371354|dbj|AP024551.1| Vibrio cholerae IDH-02387 DNA, chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2046371350|dbj|AP024547.1| Vibrio cholerae BCH-01536 DNA, chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2034570724|gb|CP047305.1| Vibrio cholerae O37 strain ATCC 25872 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2029774428|gb|CP073273.1| Vibrio fluvialis strain 19-VB00936 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2025712997|gb|CP072849.1| Vibrio cholerae MO10 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2025708070|gb|CP072847.1| Vibrio cholerae strain A1552 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2019148519|gb|CP072005.1| Vibrio fluvialis strain Colony79 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2019140445|gb|CP072007.1| Vibrio fluvialis strain Colony36 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2019136473|gb|CP072009.1| Vibrio fluvialis strain Colony42 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2019133056|gb|CP072001.1| Vibrio furnissii strain Colony85 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|2019129022|gb|CP072003.1| Vibrio furnissii strain Colony82 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|1993533143|gb|CP070474.1| Vibrio furnissii strain Colony603 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|1932138598|gb|CP064350.1| Vibrio cholerae C6706 chromosome I, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1927712664|gb|CP063420.1| Vibrio cholerae strain VC_hf7 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1890978190|gb|CP060094.1| Vibrio cholerae MO10 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|1890631614|gb|CP060047.1| Vibrio vulnificus strain 2497-87 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1877738959|gb|CP040172.1| Vibrio cholerae strain VC1374 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "\n",
      "\n",
      "G_02\n",
      "match: gi|1345471481|gb|CP014036.1| Vibrio diabolicus strain FDAARGOS_105 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|2127010576|gb|CP085843.1| Vibrio diabolicus strain NV27 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|350999931|gb|JF930483.1| Vibrio alginolyticus strain AK 1296-A2-1 actin-like cytoskeleton protein (mreB) gene, partia \n",
      "e value: 0.0264252\n",
      "match: gi|350999927|gb|JF930481.1| Vibrio alginolyticus strain LMG 11213 actin-like cytoskeleton protein (mreB) gene, partial c \n",
      "e value: 0.0264252\n",
      "\n",
      "\n",
      "G_03\n",
      "match: gi|1182626550|gb|KY771177.1| Vibrio alginolyticus strain K10M1 uridylate kinase gene, partial cds \n",
      "e value: 0.000621462\n",
      "match: gi|586909347|gb|KF899172.1| Vibrio sp. Vi_73 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.000621462\n",
      "match: gi|586909339|gb|KF899168.1| Vibrio sp. Vi_62 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.000621462\n",
      "match: gi|586909325|gb|KF899161.1| Vibrio sp. Vi_51 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.000621462\n",
      "match: gi|586909287|gb|KF899142.1| Vibrio sp. Vi_20 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.000621462\n",
      "match: gi|386799969|gb|JQ598480.1| Vibrio sp. I3K5_063 uridylate kinase (pyrH) gene, partial cds >gi|386799971|gb|JQ598481.1| V \n",
      "e value: 0.000621462\n",
      "match: gi|386799863|gb|JQ598422.1| Vibrio sp. D14K2_050 uridylate kinase (pyrH) gene, partial cds >gi|386799873|gb|JQ598427.1|  \n",
      "e value: 0.000621462\n",
      "match: gi|386799857|gb|JQ598419.1| Vibrio sp. D11K5P_093 uridylate kinase (pyrH) gene, partial cds >gi|386799859|gb|JQ598420.1| \n",
      "e value: 0.000621462\n",
      "match: gi|386143506|gb|JQ773287.1| Vibrio sp. D3K3 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.00757096\n",
      "match: gi|1785416973|gb|CP046832.1| Vibrio vulnificus strain 06-2410 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1785416962|gb|CP046823.1| Vibrio mimicus strain 06-2455 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1785416960|gb|CP046846.1| Vibrio mimicus strain F9458 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1785416959|gb|CP046820.1| Vibrio metoecus strain 2011V-1015 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1785416956|gb|CP046817.1| Vibrio metoecus strain 06-2478 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1785416950|gb|CP046825.1| Vibrio mimicus strain 07-2442 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1785416786|gb|CP046758.1| Vibrio vulnificus strain 2015AW-0208 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1776325551|gb|CP035682.1| Vibrio mimicus strain 2011V-1073 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1776325547|gb|CP035686.1| Vibrio metoecus strain 2011V-1169 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1776325545|gb|CP035688.1| Vibrio metoecus strain 08-2459 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1776325523|gb|CP035784.1| Vibrio vulnificus strain 2009V-1035 chromosome 1 \n",
      "e value: 0.0264252\n",
      "match: gi|1751004028|gb|CP044069.1| Vibrio vulnificus strain FDAARGOS_663 chromosome 2, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1736642656|gb|CP043557.1| Vibrio cholerae strain RFB05 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1375974872|gb|CP017635.1| Vibrio vulnificus Env1 chromosome I, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1337342613|gb|CP014043.2| Vibrio mimicus strain FDAARGOS_112 chromosome 2, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1320999299|gb|CP019121.1| Vibrio vulnificus strain FORC_054 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1219741010|gb|CP012881.1| Vibrio vulnificus NBRC 15645 = ATCC 27562 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1215212406|gb|CP022353.1| Vibrio sp. 2521-89 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1214590979|gb|CP014636.1| Vibrio vulnificus strain CECT 4999 chromosome I, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1209347959|gb|CP016321.1| Vibrio vulnificus strain FORC_037 chromosome I, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|1082310431|gb|CP016383.1| Vibrio mimicus strain SCCF01 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|648833925|gb|KJ604712.1| Vibrio mimicus strain XQ PyrH (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|586909431|gb|KF899214.1| Vibrio sp. Vi_78a uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|1890631614|gb|CP060047.1| Vibrio vulnificus strain 2497-87 chromosome 1, complete sequence \n",
      "e value: 0.0264252\n",
      "match: gi|159146677|gb|EU118258.1| Vibrio vulnificus strain SS108C 5C1 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|159146649|gb|EU118244.1| Vibrio vulnificus strain ATCC 43382 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|159146645|gb|EU118242.1| Vibrio mimicus strain ATCC 33653 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|159146631|gb|EU118235.1| Vibrio vulnificus strain 300 1C1 uridylate kinase (pyrH) gene, partial cds >gi|159146633|gb| \n",
      "e value: 0.0264252\n",
      "match: gi|159146629|gb|EU118234.1| Vibrio vulnificus strain 395R uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493166|gb|EF990300.1| Vibrio mimicus strain VM606 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493164|gb|EF990299.1| Vibrio mimicus strain VM605 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493162|gb|EF990298.1| Vibrio mimicus strain VM603 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493160|gb|EF990297.1| Vibrio mimicus strain VM601 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493158|gb|EF990296.1| Vibrio mimicus strain VM542 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493156|gb|EF990295.1| Vibrio mimicus strain VM535 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493154|gb|EF990294.1| Vibrio mimicus strain VM532 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493152|gb|EF990293.1| Vibrio mimicus strain VM461 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493150|gb|EF990292.1| Vibrio mimicus strain VM428 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493148|gb|EF990291.1| Vibrio mimicus strain VM343 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493146|gb|EF990290.1| Vibrio mimicus strain VM337 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "match: gi|157493144|gb|EF990289.1| Vibrio mimicus strain VM327 uridylate kinase (pyrH) gene, partial cds \n",
      "e value: 0.0264252\n",
      "\n",
      "\n",
      "G_05\n",
      "\n",
      "\n",
      "G_06\n",
      "match: gi|1864118176|gb|MT472290.1| Vibrio aestuarianus ToxR gene, partial cds \n",
      "e value: 0.000621462\n",
      "match: gi|85986463|emb|AM183573.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 02/114 \n",
      "e value: 0.000621462\n",
      "match: gi|85986461|emb|AM183572.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 02/103 \n",
      "e value: 0.000621462\n",
      "match: gi|85986459|emb|AM183571.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 02/093 \n",
      "e value: 0.000621462\n",
      "match: gi|85986457|emb|AM183570.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 02/041 \n",
      "e value: 0.000621462\n",
      "match: gi|85986451|emb|AM183567.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 01/140 \n",
      "e value: 0.000621462\n",
      "match: gi|85986449|emb|AM183566.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 01/064 \n",
      "e value: 0.000621462\n",
      "match: gi|85986445|emb|AM183564.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 01/031 >g \n",
      "e value: 0.000621462\n",
      "match: gi|85986465|emb|AM183574.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain ATCC35048 \n",
      "e value: 0.0264252\n",
      "match: gi|85986455|emb|AM183569.1| Vibrio aestuarianus partial toxR gene for transmembrane regulatory protein, strain 01/308 \n",
      "e value: 0.0264252\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"          \\n    for alignment in record.alignments:\\n        for hsp in alignment.hsps:\\n            if hsp.expect < 0.05:\\n                print('****Alignment****')\\n                #print('length:', alignment.length)\\n                print('e value:', hsp.expect)\\n                #print(hsp.query[0:75] + '   ')\\n                #print(hsp.match[0:75] + '   ')\\n                #print(hsp.sbjct[0:75] + '   ')\\n\""
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "for i in range(len(result_handle)):\n",
    "    with open('results/Osh_V1/corrected_analysis/corrected_'+str(i)+'.xml', 'w') as save_file: \n",
    "        blast_results = result_handle[i].read() \n",
    "        save_file.write(blast_results)\n",
    "'''     \n",
    "\n",
    "for i in range(6):\n",
    "    for record in NCBIXML.parse(open(\"results/vibrio_target_genes/results_second_batch_V2\"+str(i)+\".xml\")): \n",
    "         if record.alignments: \n",
    "            print(\"\\n\") \n",
    "            print(df['ID'][i])\n",
    "            #print(\"results\"+str(i))\n",
    "            for align in record.alignments: \n",
    "                for hsp in align.hsps: \n",
    "                    if hsp.expect < 0.05:\n",
    "\n",
    "                        print(\"match: %s \" % align.title[:120])\n",
    "                        print('e value:', hsp.expect)\n",
    "\n",
    "                    #print('e value:', hsp.expect)\n",
    "'''          \n",
    "    for alignment in record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if hsp.expect < 0.05:\n",
    "                print('****Alignment****')\n",
    "                #print('length:', alignment.length)\n",
    "                print('e value:', hsp.expect)\n",
    "                #print(hsp.query[0:75] + '   ')\n",
    "                #print(hsp.match[0:75] + '   ')\n",
    "                #print(hsp.sbjct[0:75] + '   ')\n",
    "'''           "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "227ee7f9",
   "metadata": {},
   "source": [
    "Put all the rsult per query in a dataframe (use the one defined above) and then find a way to characterize the species of each hit to try and filter the results.\n",
    "\n",
    "The impression is that many sequences have 16s hits.... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e0602c96",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'expect'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_17744/760922613.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      8\u001b[0m             \u001b[1;31m#print(e_val)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m             \u001b[0malign_organism\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malign\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhit_def\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 10\u001b[1;33m             \u001b[0me_val\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malign\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhsps\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexpect\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     11\u001b[0m         \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0me_val\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     12\u001b[0m         \u001b[1;31m#df.at[i,\"Matches\"] = align_organism\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'list' object has no attribute 'expect'"
     ]
    }
   ],
   "source": [
    "df[\"Matches\"] = \"\"\n",
    "\n",
    "for i in range(15):\n",
    "    for record in NCBIXML.parse(open(\"results/vibrio_16s/corrected_analysis/corrected_\"+str(i)+\".xml\")): \n",
    "        align_organism = []\n",
    "        e_val = []\n",
    "        for align in record.alignments: \n",
    "            #print(e_val)\n",
    "            align_organism.append(align.hit_def)\n",
    "            e_val.append(align.hsps.expect)\n",
    "        print(e_val)\n",
    "        #df.at[i,\"Matches\"] = align_organism\n",
    "        #df.at[i,\"e_Value\"] = e_val\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a42a0056",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Matches</th>\n",
       "      <th>e_Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G_01</td>\n",
       "      <td>ggtaacatttcaaaagcttgcttttgaa</td>\n",
       "      <td>[Shewanella algae strain ROD067 16S ribosomal ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G_02</td>\n",
       "      <td>gttgtaaagtactttcagtcgtgaggaa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G_03</td>\n",
       "      <td>gggtagaatttcaggtgtagcggtgaaa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G_04</td>\n",
       "      <td>ggggataaccattggaaacgatggctaa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G_05</td>\n",
       "      <td>tattgcacaatgggcgcaagcctgatgc</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>G_06</td>\n",
       "      <td>gtattctttgacgttagcgacagaagaa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G_07</td>\n",
       "      <td>tgatagaaatcaaggggatgcaagcgct</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>G_08</td>\n",
       "      <td>cttatgatcaatatggtcatgcagcctt</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>G_09</td>\n",
       "      <td>acattttcggtgatgtctttggtgatat</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>G_10</td>\n",
       "      <td>acgctacaacatggaattatcgctagaa</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>G_11</td>\n",
       "      <td>gattagataccctggtagtccacgccgt</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>G_12</td>\n",
       "      <td>gagcatgtggtttaattcgatgcaacgc</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>G_13</td>\n",
       "      <td>gatatccttgtttgccagcgagtaatgt</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>G_14</td>\n",
       "      <td>aatcgtggatcagaatgccacggtgaat</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>G_15</td>\n",
       "      <td>gtgggtagtttaacctttcggggaggac</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID                           Seq  \\\n",
       "0   G_01  ggtaacatttcaaaagcttgcttttgaa   \n",
       "1   G_02  gttgtaaagtactttcagtcgtgaggaa   \n",
       "2   G_03  gggtagaatttcaggtgtagcggtgaaa   \n",
       "3   G_04  ggggataaccattggaaacgatggctaa   \n",
       "4   G_05  tattgcacaatgggcgcaagcctgatgc   \n",
       "5   G_06  gtattctttgacgttagcgacagaagaa   \n",
       "6   G_07  tgatagaaatcaaggggatgcaagcgct   \n",
       "7   G_08  cttatgatcaatatggtcatgcagcctt   \n",
       "8   G_09  acattttcggtgatgtctttggtgatat   \n",
       "9   G_10  acgctacaacatggaattatcgctagaa   \n",
       "10  G_11  gattagataccctggtagtccacgccgt   \n",
       "11  G_12  gagcatgtggtttaattcgatgcaacgc   \n",
       "12  G_13  gatatccttgtttgccagcgagtaatgt   \n",
       "13  G_14  aatcgtggatcagaatgccacggtgaat   \n",
       "14  G_15  gtgggtagtttaacctttcggggaggac   \n",
       "\n",
       "                                              Matches  e_Value  \n",
       "0   [Shewanella algae strain ROD067 16S ribosomal ...      NaN  \n",
       "1                                                          NaN  \n",
       "2                                                          NaN  \n",
       "3                                                          NaN  \n",
       "4                                                          NaN  \n",
       "5                                                          NaN  \n",
       "6                                                          NaN  \n",
       "7                                                          NaN  \n",
       "8                                                          NaN  \n",
       "9                                                          NaN  \n",
       "10                                                         NaN  \n",
       "11                                                         NaN  \n",
       "12                                                         NaN  \n",
       "13                                                         NaN  \n",
       "14                                                         NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0895eb1b",
   "metadata": {},
   "source": [
    "Now the idea is for each match, take the organism (class) from the NCBI accession and then be able to plot the kind of output we got per sequence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "79e6c09c",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "esearch() missing 1 required positional argument: 'term'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_19792/2093870407.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0memail\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'ness.louafi@gmail.com'\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mhandle\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mesearch\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdb\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"nt\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0maccession\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0malign_organism\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: esearch() missing 1 required positional argument: 'term'"
     ]
    }
   ],
   "source": [
    "align_organism\n",
    "\n",
    "Entrez.email = 'ness.louafi@gmail.com'\n",
    "\n",
    "handle = Entrez.esearch(db=\"nt\",accession=align_organism[0])\n",
    "print(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50fe7a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "help(NCBIWWW.qblast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "626df2ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "<_io.StringIO object at 0x000002A56D3F01F0>\n",
      "1\n",
      "1\n",
      "<_io.StringIO object at 0x000002A56D4020D0>\n",
      "2\n",
      "2\n",
      "<_io.StringIO object at 0x000002A56D33F940>\n",
      "3\n",
      "3\n",
      "<_io.StringIO object at 0x000002A56A1C4D30>\n",
      "4\n",
      "4\n",
      "<_io.StringIO object at 0x000002A56A1F8670>\n",
      "5\n",
      "5\n",
      "<_io.StringIO object at 0x000002A56D4021F0>\n",
      "6\n",
      "6\n",
      "<_io.StringIO object at 0x000002A56D3BC3A0>\n",
      "7\n",
      "7\n",
      "<_io.StringIO object at 0x000002A56D3BC1F0>\n",
      "8\n",
      "8\n",
      "<_io.StringIO object at 0x000002A56D3BCCA0>\n",
      "9\n",
      "9\n",
      "<_io.StringIO object at 0x000002A56D65D0D0>\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from urllib.error import URLError\n",
    "import time\n",
    "\n",
    "Entrez.email = 'ness.louafi@gmail.com'\n",
    "counter = 0 \n",
    "list_record_host = []\n",
    "for record in SeqIO.parse(\"sequences/test/sequences_10.fa\", format=\"fasta\"):\n",
    "    print(record.id)\n",
    "    print(counter)\n",
    "    counter+=1\n",
    "#         print(record.seq)\n",
    "\n",
    "    # online request\n",
    "    try:\n",
    "        result_handle = NCBIWWW.qblast(\"blastn\",\"nt\",record.seq, hitlist_size = 10)\n",
    "        print(result_handle)\n",
    "    except HTTPError:\n",
    "        time.sleep(5)\n",
    "        result_handle = NCBIWWW.qblast(\"blastn\",\"nt\",record.seq, hitlist_size = 10)\n",
    "\n",
    "    # result handle stored in a list\n",
    "    list_record_host.append(result_handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "fbaec888",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(list_record_host)):\n",
    "    with open('results/results'+str(i)+'.xml', 'w') as save_file: \n",
    "        blast_results = result_handle.read() \n",
    "        save_file.write(blast_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a9399d8",
   "metadata": {},
   "source": [
    "# Save and open the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f75b2a3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for i in range(50):\n",
    "    with open('results'+str(i)+'.xml', 'w') as save_file: \n",
    "        blast_results = result_handle.read() \n",
    "        save_file.write(blast_results)\n",
    "\n",
    "E_VALUE_THRESH = 1e-20 \n",
    "for record in NCBIXML.parse(open(\"results.xml\")): \n",
    "     if record.alignments: \n",
    "        print(\"\\n\") \n",
    "        print(\"query: %s\" % record.query[:100]) \n",
    "        for align in record.alignments: \n",
    "            for hsp in align.hsps: \n",
    "                if hsp.expect < E_VALUE_THRESH: \n",
    "                    print(\"match: %s \" % align.title[:100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a33d346f",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_params = {'program': 'blastp', 'database': 'nr', 'sequence': seq_record.seq, 'expect': 10.0}\n",
    "blast_params['database'] = ['nr']\n",
    "print_data = pd.DataFrame()\n",
    "for database in blast_params['database']:\n",
    "    db_values = {}\n",
    "    result = NCBIWWW.qblast(blast_params['program'], database, blast_params['sequence'], expect=blast_params['expect'])\n",
    "    file_name = \"blast_output_\" + database + \".xml\"\n",
    "    with open(file_name, \"w\") as output_xml:\n",
    "        output_xml.write(result.read())\n",
    "    result.close()\n",
    "    result_input = open(file_name)\n",
    "    blast_records = NCBIXML.read(result_input)\n",
    "    for description in blast_records.descriptions:\n",
    "        if 'score' in db_values:\n",
    "            db_values['score'].append(description.score)\n",
    "        else:\n",
    "            db_values['score'] = [description.score]\n",
    "        if 'e-value' in db_values:\n",
    "            db_values['e-value'].append(description.e)\n",
    "        else:\n",
    "            db_values['e-value'] = [description.e]\n",
    "    df = pd.DataFrame.from_dict(db_values)\n",
    "    df['database'] = database[0:6] # we simply limit the name to the first 6 characters for easier viewing\n",
    "    frames = [print_data, df]\n",
    "    print_data = pd.concat(frames, ignore_index=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c16963af",
   "metadata": {},
   "outputs": [],
   "source": [
    "E_VALUE_THRESH = 1e-20 \n",
    "for record in NCBIXML.parse(open(\"blast_output_nr.xml\")): \n",
    "     if record.alignments: \n",
    "        print(\"\\n\") \n",
    "        print(\"query: %s\" % record.query[:100]) \n",
    "        for align in record.alignments: \n",
    "            for hsp in align.hsps: \n",
    "                if hsp.expect < E_VALUE_THRESH:\n",
    "                    if \"Arabidopsis thaliana\" in align.title:\n",
    "                        print(\"match: %s \" % align.title[:100])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49c4dd62",
   "metadata": {},
   "source": [
    "# Open when finished"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea8a8344",
   "metadata": {},
   "outputs": [],
   "source": [
    "import webbrowser\n",
    "webbrowser.open_new('http://localhost:8888/notebooks/Documents/iGEM/iGEM/Run%20BLAST.ipynb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "970026f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
